src/compute/src/compute_state.rs

```diff
@@ -164,34 +169,37 @@ impl<'a, A: Allocate, B: ComputeReplay> ActiveComputeState<'a, A, B> {
-                trace_bundle
-                    .errs_mut()
-                    .set_logical_compaction(timestamp_frontier.borrow());
-                trace_bundle
-                    .oks_mut()
-                    .set_physical_compaction(empty_frontier.borrow());
-                trace_bundle
-                    .errs_mut()
-                    .set_physical_compaction(empty_frontier.borrow());
-                // Prepare a description of the peek work to do.
-                let mut peek = PendingPeek { peek, trace_bundle };
-                // Log the receipt of the peek.
-                if let Some(logger) = self.compute_state.materialized_logger.as_mut() {
-                    logger.log(ComputeEvent::Peek(peek.as_log_event(), true));
-                }
-                // Attempt to fulfill the peek.
-                if let Some(response) = peek.seek_fulfillment(&mut Antichain::new()) {
-                    self.send_peek_response(peek, response);
-                } else {
-                    self.compute_state.pending_peeks.push(peek);
+            ComputeCommand::Peek { peek, on_replica } => {
+                // Only handle peeks that are not targeted at a different replica.
+                if on_replica.is_none() || on_replica == Some(self.compute_state.replica_id) {
+                    // Acquire a copy of the trace suitable for fulfilling the peek.
+                    let mut trace_bundle = self.compute_state.traces.get(&peek.id).unwrap().clone();
+                    let timestamp_frontier = Antichain::from_elem(peek.timestamp);
+                    let empty_frontier = Antichain::new();
+                    trace_bundle
+                        .oks_mut()
+                        .set_logical_compaction(timestamp_frontier.borrow());
+                    trace_bundle
+                        .errs_mut()
+                        .set_logical_compaction(timestamp_frontier.borrow());
+                    trace_bundle
+                        .oks_mut()
+                        .set_physical_compaction(empty_frontier.borrow());
+                    trace_bundle
+                        .errs_mut()
+                        .set_physical_compaction(empty_frontier.borrow());
+                    // Prepare a description of the peek work to do.
+                    let mut peek = PendingPeek { peek, trace_bundle };
+                    // Log the receipt of the peek.
+                    if let Some(logger) = self.compute_state.materialized_logger.as_mut() {
+                        logger.log(ComputeEvent::Peek(peek.as_log_event(), true));
+                    }
+                    // Attempt to fulfill the peek.
+                    if let Some(response) = peek.seek_fulfillment(&mut Antichain::new()) {
+                        self.send_peek_response(peek, response);
+                    } else {
+                        self.compute_state.pending_peeks.push(peek);
+                    }
```

Comment:
Not really related to this PR, so feel free to ignore, but wow would you all feel about refactoring this `match` to something like:

```rust
match cmd {
    CreateInstance { replica_id: _, logging } => self.handle_create_instance(logging),
    DropInstance => {}
    CreateDataflows(dataflows) => self.handle_create_dataflows(dataflows),
    AllowCompaction(list) => self.handle_allow_compaction(list),
    Peek { peek, on_replica } => self.handle_peek(peek, on_replica),
    CancelPeeks { uuids } => self.handle_cancel_peeks(uuids),
}
```

I.e. introducing separate handler methods for each of the match arms. I think this would make the code less cluttered and would also give us the benefit of early returns (which work with the current code as well, but only as long as nobody adds any more code after the `match`).

Reply:
Fine with me, but I'd prefer to see it in a separate PR!